staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset_test.go

```diff
@@ -32,158 +34,302 @@ import (
+		t.Errorf("Expected %#+v, got #%+v", e, a)
+	}
+	if e, a := []float64{1, 9, 7, 3}, fairAlloc([]float64{1, 9, 7, 3}, 21); !reflect.DeepEqual(e, a) {
+		t.Errorf("Expected %#+v, got #%+v", e, a)
+	}
+}
 
 type uniformClient struct {
 	hash     uint64
 	nThreads int
 	nCalls   int
 	// duration for a simulated synchronous call
 	execDuration time.Duration
-	// duration for simulated "other work"
+	// duration for simulated "other work".  This can be negative,
+	// causing a request to be launched a certain amount of time
+	// before the previous one finishes.
 	thinkDuration time.Duration
+	// When true indicates that only half the specified number of
+	// threads should run during the first half of the evaluation
+	// period
+	split bool
 }
 
-// exerciseQueueSetUniformScenario runs a scenario based on the given set of uniform clients.
+// uniformScenario describes a scenario based on the given set of uniform clients.
 // Each uniform client specifies a number of threads, each of which alternates between thinking
 // and making a synchronous request through the QueueSet.
-// This function measures how much concurrency each client got, on average, over
-// the initial evalDuration and tests to see whether they all got about the same amount.
-// Each client needs to be demanding enough to use this amount, otherwise the fair result
-// is not equal amounts and the simple test in this function would not accurately test fairness.
-// expectPass indicates whether the QueueSet is expected to be fair.
-// expectedAllRequests indicates whether all requests are expected to get dispatched.
-func exerciseQueueSetUniformScenario(t *testing.T, name string, qs fq.QueueSet, sc uniformScenario,
-	evalDuration time.Duration,
-	expectPass, expectedAllRequests, expectInqueueMetrics, expectExecutingMetrics bool,
-	rejectReason string,
-	clk *clock.FakeEventClock, counter counter.GoRoutineCounter) {
+// The test measures how much concurrency each client got, on average, over
+// the initial evalDuration and tests to see whether they all got about the fair amount.
+// Each client needs to be demanding enough to use more than its fair share,
+// or overall care needs to be taken about timing so that scheduling details
+// do not cause any client to actually request a significantly smaller share
+// than it theoretically should.
+// expectFair1 and expectFair2 indicate whether the QueueSet is expected to be
+// fair in the respective halves of a split scenario;
+// in a non-split scenario expectFair1 is what matters.
+// expectAllRequests indicates whether all requests are expected to get dispatched.
+type uniformScenario struct {
+	name                                     string
+	qs                                       fq.QueueSet
+	clients                                  []uniformClient
+	concurrencyLimit                         int
+	evalDuration                             time.Duration
+	expectFair                               []bool
```

Comment:
`[]bool expectFair` is indeed less confusing than `expectFair[1|2]`, while for now  how about we commenting something like: 

`expectFair` is an array of single element indicating whether the concurrency distribution is overall fair for non-split cases. and the on the other hand for split cases, it's a two-element array with `expectFair[0]` indicating whether the splitted half is fair and `expectFair[1]` indicating the overall distribution is fair..

Reply:
I think it is more meaningful to evaluate the two halves rather than the whole.  The whole will show a lesser degree of whatever effects there may be than an individual half.